File: crm/app/internal/ports/conversation_repository.go
// Package ports — outbound port for Conversation and Message persistence.
// Nothing here may import sqlite, http, or any adapter package.
package ports

import (
	"context"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// ConversationRepository persists and queries Conversation and Message records.
type ConversationRepository interface {
	// UpsertConversation inserts a new conversation or returns the existing one
	// matched by (channel, external_thread_id). Returns the conversation_id.
	UpsertConversation(ctx context.Context, c *domain.Conversation) (string, error)

	// GetConversationByID returns a conversation by primary key, or nil if not found.
	GetConversationByID(ctx context.Context, conversationID string) (*domain.Conversation, error)

	// UpdateConversation persists mutable fields: party_id, status, assignee_user_id,
	// last_message_at, unread_count, updated_at.
	UpdateConversation(ctx context.Context, c *domain.Conversation) error

	// InsertMessage inserts a message row; idempotent via UNIQUE(conversation_id, external_message_id).
	// Returns (messageID, inserted) — inserted=false means duplicate was skipped.
	InsertMessage(ctx context.Context, m *domain.Message) (string, bool, error)

	// ListConversations returns conversations filtered by assignee and/or status.
	// Empty string means "any" for that filter. Ordered by last_message_at DESC.
	ListConversations(ctx context.Context, assigneeUserID, status string) ([]domain.Conversation, error)

	// ListMessages returns messages for a conversation ordered by sent_at ASC.
	ListMessages(ctx context.Context, conversationID string) ([]domain.Message, error)

	// UpdateConversationOnMessage atomically advances last_message_at (take-max)
	// and increments unread_count by delta. Call only when a new message was inserted.
	// delta=1 for inbound, delta=0 for echo/outbound.
	UpdateConversationOnMessage(ctx context.Context, conversationID, sentAt string, unreadDelta int) error
}

File: crm/python/domain/ports/conversation_repository.py
from __future__ import annotations

from typing import Protocol

from domain.entities.conversation import Conversation, Message


class ConversationRepository(Protocol):
    """Outbound port for Conversation and Message persistence."""

    def upsert_conversation(self, conversation: Conversation) -> str:
        """Insert a new conversation or return the existing one matched by
        (channel, external_thread_id). Returns the conversation_id."""
        ...

    def get_conversation_by_id(self, conversation_id: str) -> Conversation | None:
        """Return a conversation by primary key, or None if not found."""
        ...

    def update_conversation(self, conversation: Conversation) -> None:
        """Persist mutable fields: party_id, status, assignee_user_id,
        last_message_at, unread_count, updated_at."""
        ...

    def insert_message(self, message: Message) -> tuple[str, bool]:
        """Insert a message row; idempotent via UNIQUE(conversation_id, external_message_id).
        Returns (message_id, inserted) — inserted=False means duplicate was skipped."""
        ...

    def list_conversations(
        self, assignee_user_id: str, status: str
    ) -> list[Conversation]:
        """Return conversations filtered by assignee and/or status.
        Empty string means 'any' for that filter. Ordered by last_message_at DESC."""
        ...

    def list_messages(self, conversation_id: str) -> list[Message]:
        """Return messages for a conversation ordered by sent_at ASC."""
        ...

    def update_conversation_on_message(
        self,
        conversation_id: str,
        sent_at: str,
        unread_delta: int,
    ) -> None:
        """Atomically advance last_message_at (take-max) and increment unread_count
        by delta. Call only when a new message was inserted.
        delta=1 for inbound, delta=0 for echo/outbound."""
        ...
